# HG changeset patch
# User John Donoghue <john.donoghue@ieee.org>
# Date 1465157851 14400
#      Sun Jun 05 16:17:31 2016 -0400
# Node ID f1f17f13b3b9f4f72d2cc545a7df0879ebc18ce0
# Parent  6780a8657be363f261bc476ab2cbc249ed498c5e
Add dialog function (Bug #48136)

* scripts/gui/dialog.m: new file

* scripts/gui/module.mk: add dialog.m to installed files

* doc/interpreter/gui.txi: add dialog to i/o dialogs help section

diff --git a/doc/interpreter/gui.txi b/doc/interpreter/gui.txi
--- a/doc/interpreter/gui.txi
+++ b/doc/interpreter/gui.txi
@@ -80,16 +80,21 @@ errors and to get textual input from the
 @DOCSTRING(msgbox)
 
 @cindex dialog, displaying a question dialog
 @DOCSTRING(questdlg)
 
 @cindex dialog, displaying a warning dialog
 @DOCSTRING(warndlg)
 
+For creating new dialog types, there is a dialog function.
+
+@cindex dialog, displaying a modal dialog
+@DOCSTRING(dialog)
+
 @node Progress Bar
 @section Progress Bar
 @cindex Progress Bar
 
 @DOCSTRING(waitbar)
 
 @node UI Elements
 @section UI Elements
diff --git a/scripts/gui/dialog.m b/scripts/gui/dialog.m
new file mode 100644
--- /dev/null
+++ b/scripts/gui/dialog.m
@@ -0,0 +1,124 @@
+## Copyright (C) 2016 John Donoghue
+##
+## This file is part of Octave.
+##
+## Octave is free software; you can redistribute it and/or modify it
+## under the terms of the GNU General Public License as published by
+## the Free Software Foundation; either version 3 of the License, or (at
+## your option) any later version.
+##
+## Octave is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+## General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with Octave; see the file COPYING.  If not, see
+## <http://www.gnu.org/licenses/>.
+
+## -*- texinfo -*-
+## @deftypefn {} {@var{h} = } dialog (@dots{}, "@var{property}", @var{value}, @dots{})
+##
+## Create an empty modal dialog window that other uicontrols can be added to.
+##
+## The dialog box is a figure object with properties as recommended for a dialog box.
+##
+## The default properties differing from a figure are:
+##
+## @table @asis
+## @item buttondownfcn
+## @code{if isempty(allchild(gcbf)), close(gcbf), end}
+##
+## @item colormap
+## []
+##
+## @item color
+## defaultuicontrolbackgroundcolor
+##
+## @item dockcontrols
+## off
+##
+## @item handlevisibility
+## callback
+##
+## @item integerhandle
+## off
+##
+## @item inverthardcopy
+## off
+##
+## @item menubar 
+## none
+##
+## @item numbertitle
+## off
+##
+## @item paperpositionmode
+## auto
+##
+## @item resize
+## off
+##
+## @item visible
+## on
+##
+## @item windowstyle
+## modal
+##
+## @end table
+## 
+##
+## Multiple property-value pairs may be specified for the dialog object, but
+## they must appear in pairs.
+##
+## The return value @var{h} is a graphics handle to the created figure.
+## object.
+##
+## Examples:
+##
+## @example
+## @group
+##
+## % create an empty dialog window titled 'Dialog Example'
+## h = dialog ("name", "Dialog Example");
+##
+## % create a button (default style)
+## b = uicontrol (h, "string", "OK", "position",[10 10 150 40], "callback","delete(gcf)");
+##
+## % wait for dialog to resume or close
+## uiwait (h);
+##
+## @end group
+## @end example
+##
+## @seealso{figure, uiwait}
+##
+## @end deftypefn
+
+## Author: jdonoghue
+
+function h = dialog (varargin)
+
+  tmph = figure ( ...
+    "buttondownfcn", "if isempty(allchild(gcbf)), close(gcbf), endif", ...
+    "color",  get(0,"defaultuicontrolbackgroundcolor"), ...
+    "colormap", [],  ...
+    "dockcontrols", "off", ...
+    "handlevisibility", "callback", ...
+    "integerhandle", "off", ...
+    "inverthardcopy", "off", ...
+    "menubar", "none", ...
+    "numbertitle", "off", ...
+    "paperpositionmode", "auto", ...
+    "resize", "off", ...
+    "toolbar", "none", ...
+    "visible", "on", 
+    "windowstyle", "modal", ...
+    varargin{:} );
+    
+  if (nargout > 0)
+    h = tmph;
+  endif 
+endfunction
+
+
diff --git a/scripts/gui/module.mk b/scripts/gui/module.mk
--- a/scripts/gui/module.mk
+++ b/scripts/gui/module.mk
@@ -8,16 +8,17 @@ scripts_gui_PRIVATE_FCN_FILES = \
   scripts/gui/private/__get_funcname__.m \
   scripts/gui/private/__is_function__.m \
   scripts/gui/private/__uigetdir_fltk__.m \
   scripts/gui/private/__uigetfile_fltk__.m \
   scripts/gui/private/__uiobject_split_args__.m \
   scripts/gui/private/__uiputfile_fltk__.m
 
 scripts_gui_FCN_FILES = \
+  scripts/gui/dialog.m \
   scripts/gui/errordlg.m \
   scripts/gui/guidata.m \
   scripts/gui/guihandles.m \
   scripts/gui/helpdlg.m \
   scripts/gui/inputdlg.m \
   scripts/gui/listdlg.m \
   scripts/gui/msgbox.m \
   scripts/gui/questdlg.m \
